control-plane/agents/src/bin/core/tests/rebuild/mod.rs

```diff
@@ -214,4 +300,213 @@ async fn replica_wait_and_full_rebuild() {
+            ),
+        )
+        .await
+        .unwrap();
+
+    let volume_state = volume.state.clone();
+    assert!(volume_state.status == VolumeStatus::Online);
+    let nexus = volume_state.target.unwrap();
+    assert_eq!(nexus.children.len(), 3);
+    nexus
+        .children
+        .iter()
+        .for_each(|c| assert!(c.state == ChildState::Online));
+
+    // Check where the replicas are, apart from vol target node.
+    let replicas = api_client.replicas_api().get_replicas().await.unwrap();
+    tracing::info!({?replicas}, "Available replicas - ");
+    let testrep = replicas.iter().find(|r| r.node != vol_target).unwrap();
+    let testrep_node = &testrep.node;
+    tracing::info!(
+        "Restarting node {testrep_node} having replica {}",
+        testrep.uri.clone()
+    );
+
+    // Restart the container.
+    cluster
+        .composer()
+        .restart(&testrep_node.to_string())
+        .await
+        .expect("container stop failure");
+
+    // Check the pool on restarted node is imported back.
+    wait_till_pool_online(&cluster, testrep, Duration::from_secs(20)).await;
+
+    // Fetch replicas again.
+    let replicas_now = api_client.replicas_api().get_replicas().await.unwrap();
+    // The replica picked for test must be online now again.
+    assert!(replicas_now.iter().any(|r| r.uri == testrep.uri));
+    if let Some(rep) = replicas_now.iter().find(|r| r.uri == testrep.uri) {
+        assert_eq!(rep.state, ReplicaState::Online);
+    }
+
+    // Get the volume again for validations.
+    let vol = volume_api.get_volume(volid.uuid()).await.unwrap();
+    let volume_state = vol.state;
+    let nexus = volume_state.target.unwrap();
+    // The child must be still in the nexus.
+    let is_removed = !nexus.children.iter().any(|c| c.uri == testrep.uri);
+    assert!(!is_removed);
+    // TODO: We can additionally check that the volume state eventually becomes Online,
+    // and doesn't remain Degraded once the partial rebuild finishes. Currently no support for
+    // getting rebuild info via nexus grpc client NexusOperations.
+
+    let nexus_client = cluster.grpc_client().nexus();
+    let nexus_id = NexusId::from(volume.state.target.unwrap().uuid);
+
+    wait_nexus_online(&nexus_client, nexus_id.clone())
+        .await
+        .expect("Rebuild didnt finish in 30 secs");
+
+    let history = nexus_client
+        .get_rebuild_history(&GetRebuildRecord::new(nexus_id.clone()), None)
+        .await
+        .expect("Failed to get rebuild record");
+
+    assert_eq!(
+        nexus_id, history.uuid,
+        "Cant match nexus id in rebuild history"
+    );
+
+    assert_eq!(
+        1,
+        history.records.len(),
+        "Number of rebuild history is not equal to 1"
+    );
+
+    assert!(
+        history.records.get(0).unwrap().is_partial,
+        "Rebuild type is not Partial rebuild"
+    )
+}
+
+const POOL_SIZE_BYTES: u64 = 128 * 1024 * 1024;
+#[tokio::test]
+async fn validate_empty_rebuild_record() {
```

Comment:
We're adding a lot of new clusters here, let's try to reuse clusters to keep  test runs down..

Reply:
This i can test in existing test before rebuild starts.

Reply:
checking empty history for nexus in existing rebuild history case. We have 4 cases now.